### PR TITLE
Add support for constructors, and integration tests.

### PR DESCRIPTION
**Pull request by emilio**
r? @fitzgen 

cc @vvuk

### CODE REVIEW

File: libbindgen/tests/headers/constructors.hpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+
+class TestOverload {
+  // This one shouldnt' be generated.
```
Comment (fitzgen) [inline libbindgen/tests/headers/constructors.hpp]: Because it is private, right?

File: libbindgen/tests/headers/constructors.hpp [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  TestOverload();
+public:
+  TestOverload(int);
+  TestOverload(double);
```
Comment (fitzgen) [inline libbindgen/tests/headers/constructors.hpp]: Do we handle when the class already has a `new` method properly? Do we just treat the constructors or the `new` methods as overloads? Care to add a small test case?

Comment (fitzgen) [inline libbindgen/tests/headers/constructors.hpp]: Oh wait -- `new` is a keyword ofc -- nvm

**Comment by vvuk** [comment]
Works great with Skia, fwiw.  Thanks!

I'll file a feature request -- would be nice to be able to rename the rust-side constructors based on their signature by passing in some kind of mapping file to bindgen, otherwise there's no guarantee that the function names will be stable (e.g. if upstream adds another constructor/overload).

File: libbindgen/src/ir/function.rs [hunk 3]
```diff
@@ -147,6 +147,7 @@ impl FunctionSig {
         };
         let mut args: Vec<_> = match cursor.kind() {
             CXCursor_FunctionDecl |
+            CXCursor_Constructor |
             CXCursor_CXXMethod => {
```
Comment (fitzgen) [inline libbindgen/src/ir/function.rs]: Is there a reason this doesn't also match `CXCursor_Destructor` like the `ClangSubItemParser` impl further down does?

Is it handled elsewhere?

File: libbindgen/src/ir/function.rs [hunk 4]
```diff
@@ -184,10 +185,12 @@ impl FunctionSig {
             }
         };
 
-        if cursor.kind() == CXCursor_CXXMethod {
-            let is_const = cursor.method_is_const();
-            let is_virtual = cursor.method_is_virtual();
-            let is_static = cursor.method_is_static();
+        let is_method = cursor.kind() == CXCursor_CXXMethod;
+
+        if is_method || cursor.kind() == CXCursor_Constructor {
```
Comment (fitzgen) [inline libbindgen/src/ir/function.rs]: Ditto about dtors for this check.

File: libbindgen/src/ir/function.rs (hunk 3 above)
Comment (fitzgen) [inline libbindgen/src/ir/function.rs]: Ah -- are we not binding destructors yet? That's surprising I really thought we were.

File: libbindgen/src/codegen/mod.rs [hunk 5]
```diff
@@ -1359,13 +1359,14 @@ impl MethodCodegen for Method {
         let mut stmts = vec![];
 
         // If it's a constructor, we need to insert an extra parameter with a
-        // variable called `tmp` we're going to create.
+        // variable called `__bindgen_tmp` we're going to create.
```
Comment (fitzgen) [inline libbindgen/src/codegen/mod.rs]: Great :)

File: bindgen-integration/cpp/Test.h [hunk 6]
```diff
@@ -1,6 +1,6 @@
 #pragma once
 
-class Test final {
```
Comment (fitzgen) [inline bindgen-integration/cpp/Test.h]: Why no more final? We can't get at it with `llvm_stable`?

File: .travis.yml [hunk 7]
```diff
@@ -38,7 +38,8 @@ script:
   - cargo test --features "$BINDGEN_FEATURES"
   - cargo test --release --features "$BINDGEN_FEATURES"
   - cd ../bindgen-integration
-  - cargo test
+  - cargo test --features "$BINDGEN_FEATURES"
+  - cargo test --release --features "$BINDGEN_FEATURES"
```
Comment (fitzgen) [inline .travis.yml]: I think this is getting big enough that we should pull it out into a `./ci/test` script. In a follow up, of course.

File: libbindgen/src/ir/comp.rs [hunk 8]
```diff
@@ -466,6 +466,7 @@ impl CompInfo {
         &self.methods
     }
 
+    /// Get this type's set of constructors.
```
Comment (fitzgen) [inline libbindgen/src/ir/comp.rs]: Gotcha ;)

File: libbindgen/tests/headers/constructors.hpp (hunk 1 above)
Comment (emilio) [inline libbindgen/tests/headers/constructors.hpp]: Right.

File: bindgen-integration/cpp/Test.h (hunk 6 above)
Comment (emilio) [inline bindgen-integration/cpp/Test.h]: The travis gcc was complaining, and I didn't want to fight that battle :)

File: libbindgen/src/ir/function.rs (hunk 3 above)
Comment (emilio) [inline libbindgen/src/ir/function.rs]: Yeah, we're not doing it. We can without too much trouble though.

When we do, there's no point in doing that, though (we could, I guess?). But destructors have no arguments.

File: libbindgen/src/ir/function.rs (hunk 4 above)
Comment (emilio) [inline libbindgen/src/ir/function.rs]: Yeah, when we support them that check should be there. But we should also check for virtual destructors, etc.

**Pull request merged**